### ISSUE TITLE
Could abstract algorithm use types to rewrite the typed term?

### ISSUE THREAD

**Issue by Josef-Vonasek** [MaiaVictor/abstract-algorithm#11]
Lets take fusion as an example:

1. this fuses
     ```elm
    @List.fmap #f /Y #fmap #xs #con #nil
      //xs
        #x #xs //con /f x /fmap xs
        nil
     ```
2. this does not
    ```elm
    @List.map #f /Y #map #xs
      //xs
        #x #xs #con #nil //con /f x /map xs
        #con #nil nil
     ```

Obviously it is not possible to transform `map` to `fmap`, unless we know that `xs` is *well behaved*. (Bad behaved `xs` would be for example `#con #nil #x x`.)

-----

Now comes the tough question? How do we know `xs` is well behaved? 

* In pure lambda calculus it's impossible. 
* Haskell has ADT and `case of` which allows this particular transformation.
* With (dependent) type system it should be doable, but nobody has done it AFAIK, so it's probably nontrivial.

*Could we do it?*

**Comment by Josef-Vonasek** [comment]
To make the task simpler you might think of annotating each argument with set of all possible input values:

```elm
@N        #c #n n
@C #x #xs #c #n //c x xs
@List.map 
  #(f : Any) 
  /Y #(map: List.map) -- this is recursive definition :/
     #(xs: C Any List | N)
      //xs
        ...
 ```
**The naive solution** would be then to evaluate `map` with all possible arguments (we have them listed in the annotations):

```elm
@List.map #f /Y #map #xs
    #con #nil nil

@List.map #f /Y #map #xs
    #con #nil //con /f Any /map List
```
now it is obvious map always returns lambda starting with `#con #nil ...` so `map` becomes `fmap`.

----

This approach has sadly exponential complexity, does not involve interaction nets and is quite **ad hoc**.

But I think it could point us in the right direction.

**Comment by MaiaVictor** [comment]
Do you have any paper describing a type system like that one you described? I don't think i've seen a type system which works by simply listing all possible values of its inputs, looks interesting.

**Comment by MaiaVictor** [comment]
But anyway, before answering that question, I think we must understand exactly when and why fusion happens, it is not as simple as that. Some terms require slightly different strategies. All in all it is still very new and unpredictable to me.

**Comment by Josef-Vonasek** [comment]
> Do you have any paper describing a type system like that one you described? I don't think i've seen a type system which works by simply listing all possible values of its inputs, looks interesting.

No, it is just an experiment. But it could be emulated by doing runtime checks for equality on each function application which is not very efficient, but some languages do it (e.g. python).

**Comment by Josef-Vonasek** [comment]
Then we should at least make a file, where we put all before => after transformations, which allowed fusion, so we can make some sense of it.

**Comment by Josef-Vonasek** [comment]
For example I still do not understand, why does `fmap` require `identity` to not loop forever. Do you have any explanation for it?

**Comment by MaiaVictor** [comment]
I agree with a before => after file. And I have no idea too, a graphical evaluator would be handy to understand why.

**Comment by Josef-Vonasek** [comment]
As an example in [theorems for free](http://ecee.colorado.edu/ecen5533/fall11/reading/free.pdf)  polymorphism is used to derive theorems about terms.
